migration/src/bin/migrate_to.rs

```diff
@@ -0,0 +1,604 @@
+    #[arg(value_parser = available_migrations())]
+    target_version: String,
+    #[arg(short, long)]
+    dry_run: bool,
+    #[arg(short = 'u', long, env = "DATABASE_URL")]
+    database_url: String,
+}
+
+#[async_std::main]
+async fn main() -> Result<(), Error> {
+    let args = Args::parse();
+    tracing_subscriber::fmt()
+        .with_env_filter(EnvFilter::from_default_env())
+        .init();
+
+    let db = Database::connect(ConnectOptions::new(args.database_url)).await?;
+    check_database_is_compatible::<Migrator>(&db).await?;
+    match args.direction {
+        Direction::Up => migrate_up::<Migrator>(&db, args.dry_run, &args.target_version).await?,
+        Direction::Down => {
+            migrate_down::<Migrator>(&db, args.dry_run, &args.target_version).await?
+        }
+    }
+    Ok(())
+}
+
+// Applied migrations must be a subset of the migrations available by the
+// MigratorTrait for this CLI to be operated safely.
+async fn check_database_is_compatible<M: MigratorTrait>(
+    db: &DatabaseConnection,
+) -> Result<(), Error> {
+    // If the database is uninitialized, we can continue. The migrator will
+    // initialize the database for us.
+    if !has_migrations_table(db).await? {
+        return Ok(());
+    }
+
+    let migrations: Vec<String> = M::migrations().iter().map(|m| m.name().into()).collect();
+    let applied_migrations = applied_migrations(db).await?;
+
+    let print_error = || {
+        error!("expected migrations: {:?}", migrations);
+        error!("present migrations: {:?}", applied_migrations);
+    };
+    for (i, applied) in applied_migrations.iter().enumerate() {
+        match migrations.get(i) {
+            Some(migration) if migration != applied => {
+                print_error();
+                return Err(Error::DbNotCompatible);
+            }
+            None => {
+                print_error();
+                return Err(Error::DbNotCompatible);
+            }
+            _ => {}
+        }
+    }
+    Ok(())
+}
+
+async fn migrate_up<M: MigratorTrait>(
+    db: &DatabaseConnection,
+    dry_run: bool,
+    target: &str,
+) -> Result<(), Error> {
+    let target_index =
+        migration_index::<M>(target).ok_or(Error::MigrationNotFound(target.to_string()))?;
+
+    let migrations_range;
+    let num_migrations = match latest_applied_migration(db).await {
+        Ok(latest_migration) => {
+            let latest_index =
+                migration_index::<M>(&latest_migration).ok_or(Error::DbMigrationNotFound)?;
+            match target_index.cmp(&latest_index) {
+                Ordering::Less => return Err(Error::VersionTooOld(target.to_string())),
+                Ordering::Equal => {
+                    info!("no action taken, already at desired version");
+                    return Ok(());
+                }
+                Ordering::Greater => {
+                    migrations_range = (latest_index + 1)..=target_index;
+                    target_index - latest_index
```

Comment:
nit: assigning to a tuple is IMO a little more idiomatic.
```suggestion
    let (migrations_range, num_migrations) = match latest_applied_migration(db).await {
        Ok(latest_migration) => {
            let latest_index =
                migration_index::<M>(&latest_migration).ok_or(Error::DbMigrationNotFound)?;
            match target_index.cmp(&latest_index) {
                Ordering::Less => return Err(Error::VersionTooOld(target.to_string())),
                Ordering::Equal => {
                    info!("no action taken, already at desired version");
                    return Ok(());
                }
                Ordering::Greater => {
                    ((latest_index + 1)..=target_index, target_index - latest_index)
```
... and equivalently in the `Err(err) if matches!(...)` arm below